File: api/src/models/item.py
"""
Item models

Pydantic models for the items API.
"""

from typing import Optional
from pydantic import BaseModel, Field


class ItemCreate(BaseModel):
    """Request model for creating an item."""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    quantity: int = Field(1, ge=1)


class ItemUpdate(BaseModel):
    """Request model for updating an item."""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    quantity: Optional[int] = Field(None, ge=1)


class BubbleRef(BaseModel):
    """Reference to a bubble."""
    id: str
    name: str


class Item(BaseModel):
    """Response model for an item."""
    id: str
    owner_id: str
    name: str
    description: Optional[str] = None
    quantity: int = 1
    created_at: str


class ItemWithShares(Item):
    """Item with shared bubbles info."""
    shared_bubbles: list[BubbleRef] = []
    available_quantity: int = 0  # quantity minus active loans


class ItemList(BaseModel):
    """Paginated list of items."""
    items: list[ItemWithShares]
    total: int
    page: int
    limit: int


class ItemShare(BaseModel):
    """Request model for sharing an item to bubbles."""
    bubble_ids: list[str]

File: api-rust/src/models/item.rs
//! Item models

use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

/// Request model for creating/updating an item
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct ItemCreate {
    /// Item name
    pub name: String,
    /// Optional description
    pub description: Option<String>,
}

/// Item response model
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct Item {
    /// Unique identifier
    pub id: String,
    /// Item name
    pub name: String,
    /// Optional description
    pub description: Option<String>,
    /// Creation timestamp
    pub created_at: String,
    /// Owner user ID
    pub user_id: String,
}

/// Database model (includes updated_at)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ItemDb {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub user_id: String,
}

impl From<ItemDb> for Item {
    fn from(db: ItemDb) -> Self {
        Item {
            id: db.id,
            name: db.name,
            description: db.description,
            created_at: db.created_at,
            user_id: db.user_id,
        }
    }
}

/// Paginated list of items
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct ItemList {
    /// List of items
    pub items: Vec<Item>,
    /// Total count
    pub total: i64,
    /// Current page
    pub page: i32,
    /// Items per page
    pub limit: i32,
}
